src/NzbDrone.Core/Messaging/Commands/CommandService.cs

```diff
@@ -0,0 +1,193 @@
+        void SetMessage(CommandModel command, string message);
+        void Completed(CommandModel command);
+        void Failed(CommandModel command, Exception e);
+    }
+
+    public class CommandService : ICommandService, IHandle<ApplicationStartedEvent>
+    {
+        private readonly ICommandRepository _repo;
+        private readonly IServiceFactory _serviceFactory;
+        private readonly IEventAggregator _eventAggregator;
+        private readonly Logger _logger;
+
+        private ICached<string> _messageCache; 
+
+        private static readonly object Mutex = new object();
+
+        public CommandService(ICommandRepository repo, 
+                              IServiceFactory serviceFactory,
+                              IEventAggregator eventAggregator,
+                              ICacheManager cacheManager,
+                              Logger logger)
+        {
+            _repo = repo;
+            _serviceFactory = serviceFactory;
+            _eventAggregator = eventAggregator;
+            _logger = logger;
+
+            _messageCache = cacheManager.GetCache<string>(GetType());
+        }
+
+        public CommandModel PublishCommand<TCommand>(TCommand command) where TCommand : Command
+        {
+            Ensure.That(command, () => command).IsNotNull();
+
+            _logger.Trace("Publishing {0}", command.GetType().Name);
+
+            lock (Mutex)
+            {
+                var existingCommands = _repo.FindCommands(command.Name).Where(c => c.Status == CommandStatus.Queued ||
+                                                                                   c.Status == CommandStatus.Started)
+                                            .ToList();
+
+                var existing = existingCommands.SingleOrDefault(c => CommandEqualityComparer.Instance.Equals(c.Body, command));
+
+                if (existing != null)
+                {
+                    _logger.Trace("Command is already in progress: {0}", command.GetType().Name);
+
+                    return existing;
+                }
+
+                var commandModel = new CommandModel
+                                   {
+                                       Name = command.Name,
+                                       Body = command,
+                                       Queued = DateTime.UtcNow,
+                                       Trigger = command.Trigger,
+                                       Priority = CommandPriority.Normal,
+                                       Status = CommandStatus.Queued
+                                   };
+
+                _repo.Insert(commandModel);
+
+                return commandModel;
+            }
+        }
+
+        public CommandModel PublishCommand(string commandName)
+        {
+            var command = PublishCommand(commandName, null);
+
+            return command;
+        }
+
+        public void PublishScheduledTasks(List<ScheduledTask> scheduledTasks)
```

Comment:
I don't think this belongs here. Should be part of the scheduler or whatever actually makes this call.

CommandService should not care where stuff comes from, it just adds stuff to the queue.